pkg/volume/flexvolume/probe.go

```diff
@@ -127,46 +157,93 @@ func (prober *flexVolumeProber) handleWatchEvent(event fsnotify.Event) error {
 	if err != nil {
 		return err
 	}
-
+	parentPathAbs := filepath.Dir(eventPathAbs)
 	pluginDirAbs, err := filepath.Abs(prober.pluginDir)
 	if err != nil {
 		return err
 	}
 
-	// If the Flexvolume plugin directory is removed, need to recreate it
-	// in order to keep it under watch.
-	if eventOpIs(event, fsnotify.Remove) && eventPathAbs == pluginDirAbs {
-		if err := prober.createPluginDir(); err != nil {
-			return err
+	// event of pluginDirAbs
+	if eventPathAbs == pluginDirAbs {
+		// If the Flexvolume plugin directory is removed, need to recreate it
+		// in order to keep it under watch.
+		if eventOpIs(event, fsnotify.Remove) {
+			if err := prober.createPluginDir(); err != nil {
+				return err
+			}
+			if err := prober.addWatchRecursive(pluginDirAbs); err != nil {
+				return err
+			}
 		}
-		if err := prober.addWatchRecursive(pluginDirAbs); err != nil {
-			return err
+		return nil
+	}
+
+	// event of subdir of pluginDirAbs
+	if parentPathAbs == pluginDirAbs {
+		// dir is created in pluginDirAbs directly
+		if eventOpIs(event, fsnotify.Create) {
+			if err := prober.addWatchRecursive(eventPathAbs); err != nil {
+				return err
+			}
 		}
-	} else if eventOpIs(event, fsnotify.Create) {
+		return nil
+	}
+
+	// watch newly added subdirectories inside a driver directory
+	if eventOpIs(event, fsnotify.Create) {
 		if err := prober.addWatchRecursive(eventPathAbs); err != nil {
 			return err
 		}
 	}
 
-	prober.updateProbeNeeded()
+	eventRelPathToPluginDir, err := filepath.Rel(pluginDirAbs, eventPathAbs)
+	if err != nil {
+		return err
+	}
+
+	// event inside specific driver dir
+	if len(eventRelPathToPluginDir) > 0 {
+		driverDirName := topDirName(eventRelPathToPluginDir)
+		driverDirAbs := filepath.Join(pluginDirAbs, driverDirName)
+		if len(driverDirAbs) > 0 {
+			parts := strings.Split(driverDirName, "~")
```

Comment:
nit: Could you separate the logic to compute the path to the executable in a separate helper function? Easier to read that way

Reply:
done!